Move the strange refunding logic from Spree::TaxRate into the default tax calculator

mamhoff opened the pull request:
When reviewing #904, we realized that we can not change the logic of
Spree::TaxRate#compute_amount if we want the migration included there to
always to the right thing. Because that method modifies the tax amount of things,
it should actually be in a calculator. I now moved that logic to the default tax
calculator, with the intention of deprecating use of that calculator entirely.

There is code for pre-Spree-2.2 order-based taxation in there, as well as code for
generating tax refunds, which in the future we won't need either.

gmacdougall commented:
I'd like to see the unit tests for default_tax.rb improved, covering the new functionality and providing better tests for the existing logic.

We also need to make the default_tax handle currencies that do not have two decimal places, but I'm fine with leaving that for another PR.

mamhoff commented:
Currencies with more than two decimal places are not supported at all in Solidus, I think that's an entirely different issue. And yeah, I'll provide unit tests for default tax.

gmacdougall commented:
I was thinking more currencies without decimal places like JPY or KRW which are generally supported, aside from some bad rounding logic here and there.

mamhoff commented:
Ha. Yeah, I hear you.

mamhoff commented:
Refactored the specs and added specs for some of the refunding logic @gmacdougall

mamhoff commented:
The failing spec on 6990e65 is due to a PhantomJS crash, I'm quite sure.

gmacdougall commented on core/spec/models/spree/calculator/default_tax_spec.rb:
```diff
@@ -70,51 +68,145 @@
+    let(:line_item) { Spree::LineItem.new }
+
+    it 'calls #compute_line_item' do
+      expect(subject).to receive(:compute_line_item)
```
I don't like stubbing methods on the file that we're writing the spec for. I don't think that this test provides value.

mamhoff commented:
Rebased and added specs for the actual behaviour of the calculator when called with line item, shipment or shipping rate. I think this is good to go now. @gmacdougall @jhawthorn

adammathys commented:
:+1: Looks good to me.

mamhoff commented:
@gmacdougall If you have a minute, this should be good to go.

mamhoff commented:
Please please review this, it's base work for #904, which is also ripe.

jhawthorn commented:
:+1: Looks good to me

gmacdougall commented:
:+1:
